### ISSUE TITLE
[MacOS] Empty `std::vector<VideoMode>` returned from `getFullscreenModes()`, causing segfault

### ISSUE THREAD

**Issue by md81544** [SFML/SFML#2300]
## Subject of the issue

This issue is described in the forums, here:
https://en.sfml-dev.org/forums/index.php?topic=28601.0

I am not the OP of that thread, but I (and others) are seeing the same issue. Essentially, on MacOS (latest versions), a call to `getFullScreenModes()` (https://github.com/SFML/SFML/blob/2.6.x/src/SFML/Window/OSX/VideoModeImpl.cpp#L40) returns an empty vector of video modes. This is because each candidate is rejected by the `continue` at line 65, owing to all of them being larger than the reported desktop width/height.

Adding some debug output at that point shows the following modes are all rejected:

```
mode.size.x = 3840, desktop.size.x = 3420, mode.size.y = 2400, desktop.size.y = 2224
mode.size.x = 4096, desktop.size.x = 3420, mode.size.y = 2560, desktop.size.y = 2224
mode.size.x = 4096, desktop.size.x = 3420, mode.size.y = 2664, desktop.size.y = 2224
mode.size.x = 5120, desktop.size.x = 3420, mode.size.y = 3200, desktop.size.y = 2224
mode.size.x = 5120, desktop.size.x = 3420, mode.size.y = 3328, desktop.size.y = 2224
(Empty std::vector of modes returned)
```

This appears to be related to the question about whether Apple's `CGDisplayModeGetWidth()` function returns its value in points or pixels.

## Your environment

* OS:  MacOS 13.1 "Ventura"
* Version of SFML: I've tried 3.0 and 2.5.x branches
* Apple clang version 14.0.0 (clang-1400.0.29.202)
* Special compiler flags used: just the usual `-std=c++17 -g -Wall -Wextra -Werror -Wpedantic`

## Steps to reproduce

This is using SFML 3.0 but a similar program for 2.5.x exhibits the same behaviour. Note, this issue is on MacOS only.

```cpp
#include <SFML/Graphics.hpp>
#include <SFML/Window.hpp>

#include <iostream>
#include <vector>

int main()
{
    std::vector<sf::VideoMode> modes = sf::VideoMode::getFullscreenModes();
    if (modes.empty())
    {
        std::cout << "Empty set of modes returned\n";
    }
    else
    {
        for (std::size_t i = 0; i < modes.size(); ++i)
        {
            sf::VideoMode mode = modes[i];
            std::cout << "Mode #" << i << ": " << mode.size.x << "x" << mode.size.y << " - " << mode.bitsPerPixel
                      << " bpp" << std::endl;
        }
    }

    sf::RenderWindow win(sf::VideoMode(sf::Vector2u{800, 600}), "Test", sf::Style::Fullscreen);
    return 0;
}

```

## Expected behavior

We should not see a segfault, but SFML is assuming it will always have a valid mode as a default (after it states "The requested video mode is not available, switching to a valid mode") - but `getFullscreenModes()` is returning an empty vector, so it segfaults.

## Actual behavior

Segfault owing to the empty vector.

Do please let me know if anything needs clarification.

**Comment by JonnyPtn** [comment]
What’s the actual physical resolution of your screen?

**Comment by md81544** [comment]
2560x1664 is the native resolution of mine, which is an M2 Macbook Air (https://www.apple.com/uk/macbook-air-m2/specs/)

**Comment by Rivera-Josh** [comment]
We're having the same issue on an M1 MacBook Pro 16 inch. (3456x2234). We can't get any fullscreen mode to work at all. Is there any extra information that we can provide to help fix this?

**Comment by ChrisThrasher** [comment]
I can reproduce this on my M1 MacBook Pro. Sorry I don't have any ideas for a fix but I'm happy to work with someone who may have the time or skills to take a deeper look into this.

**Comment by JonnyPtn** [comment]
First step would be to remove that condition and see how the resulting modes work. Looks like it was added in #1042 (7 years ago) but not immediately apparent why. If removing that check solves the issue then we can work backwards on the older versions (if needed) to see what it breaks, if anything.

**Comment by md81544** [comment]
Removing the check definitely works in my testing. There's still the issue with the reported sizes of the available modes being much larger than the physical resolution.

**Comment by JonnyPtn** [comment]
So that’s likely an issue with DPI scaling which SFML doesn’t really handle at this point - does it render at the larger resolution, meaning it’s cut-off by the screen limits? Could you try fiddling with the “high resolution capable” property in your apps info.plist to see how that affects it?

**Comment by md81544** [comment]
I did experiment with the info.plist path when I was investigating initially. I couldn't get anything I tried to work. It could be me, I am a complete newb when it comes to Mac development : )

wrt your first question, about does it render at the larger resolution, if I display a `[CircleShape](sf::CircleShape  shape(555.f);)` then it fills the entire height of the display, which implies that even though I constructed the window with `sf::VideoMode(sf::Vector2u{5120, 3328})`, the actual y-resolution is around 1100px.

**Comment by coinop-logan** [comment]
That's funny. When I build my project on friends' macs, I get the opposite problem of @md81544: everything I render is smaller than it "should be". In other words if I draw a rectangle with width exactly equal to the *reported* width of the window I create, it only fills up half or less of the screen.

**Comment by coinop-logan** [comment]
I've read and re-read both this thread and the original forum thread a few times, while trying to find any solution or workaround for my project. I've also switched between SFML 2.5 and 3, as well as Mac M1 and non-M1 macs.

Here are some summaries/thoughts that occur to me so far. Some of this is deduction on my part, so please correct anything here that seems wrong!

* A possible workaround, it sounds like, is to neuter one or two lines of code from SFML. From the forum thread:
  * Removing [this continue](https://github.com/SFML/SFML/blob/2.6.x/src/SFML/Window/OSX/VideoModeImpl.cpp#L65) allows the apparently-too-large-and-thus-invalid modes through, resulting in a non-empty vector being returned. This allows SFML to continue without crashing, though it still then resorts to a backup mode instead of whatever the user selects from that list.
  * [This line](https://github.com/SFML/SFML/blob/82b48a7520a2eb83576ec16d68ff9ea6d3d5d5bd/src/SFML/Window/OSX/cg_sf_conversion.mm#L94) seems to be responsible for the too-large resolutions above. Disabling this line seems to also work around the problem, at least insofar as making sure the returned vector is non-empty and that its values are valid. It's not clear whether this actually works as far as giving the user the resolution he selects from the vector. Maybe someone could test this?
* In the forum someone mentions [this StackOverflow thread](https://stackoverflow.com/questions/42024942/cgdisplaymodegetwidth-height-sometimes-returns-pixels-sometimes-points) regarding `CGDisplayModeGetWidth`/`Height`'s inconsistent behavior and documentation. At first glance, it seems very likely to be related to this issue, if not a direct cause. But then if you read the full conversation in the thread, it's implied that the OP of that question was just mistaken about his testing setup so... I don't really know what to make of this - could be something, could be nothing.

And this might be unrelated but...

* I was running into an issue that smells very related to this, on SFML 2.5 and non-M1 macs (wasn't able to test at all with M1 macs until I upgraded to SFML 3.0). In these cases, the game would open in fullscreen mode, but SFML would end up "confused" about what the resolution was. It seemed to be that Mac never changed from its native resolution, but SFML thought it had. The result is that the game would render in a limited-size portion of the full screen, the size of the portion corresponding to the user's chosen resolution. So if you tried a 1920x resolution on a Mac that had a ~3300x resolution, your full screen would be "captured" by the app but the game itself would still be rendered very small.
  * In these cases, I've tried but failed to find a way to simply detect what the actual resolution of the game is after the window is up, hoping that at the very least I could get the real situation and deal with it in code. So far I have not found a way to do that.

**Comment by ChrisThrasher** [comment]
> [This line](https://github.com/SFML/SFML/blob/82b48a7520a2eb83576ec16d68ff9ea6d3d5d5bd/src/SFML/Window/OSX/cg_sf_conversion.mm#L94) seems to be responsible for the too-large resolutions above. Disabling this line seems to also work around the problem, at least insofar as making sure the returned vector is non-empty and that its values are valid. It's not clear whether this actually works as far as giving the user the resolution he selects from the vector. Maybe someone could test this?

This line appears to convert values in points to values in pixels, is it not? I ask because that Stack Overflow thread mentions `CGDisplayModeGetPixel{Width|Height}` which were added in macOS 10.8. If we want points then we should be able to use `CGDisplayModeGet{Width|Height}` and use `CGDisplayModeGetPixel{Width|Height}` if we want pixels. If both APIs exist then I'm not sure why we need to multiple by a scaling factor to get the correct result.

**Comment by ChrisThrasher** [comment]
```cpp
////////////////////////////////////////////////////////////
VideoMode convertCGModeToSFMode(CGDisplayModeRef cgmode)
{
    // The main documentation says the sizes returned by
    // CGDisplayModeGetWidth and CGDisplayModeGetHeight
    // are expressed in pixels. However, some additional
    // documentation [1] states they actually return
    // values in points starting with 10.8. In the same
    // release CGDisplayModeGetPixel{Width|Height} was
    // added which directly returns the size in pixels.
    //
    // [1]: "APIs for Supporting High Resolution" > "Additions and Changes for OS X v10.8"
    // https://developer.apple.com/library/mac/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/APIs/APIs.html#//apple_ref/doc/uid/TP40012302-CH5-SW27
    return VideoMode(Vector2u(Vector2(CGDisplayModeGetPixelWidth(cgmode), CGDisplayModeGetPixelHeight(cgmode))),
                     modeBitsPerPixel(cgmode));
}
```

I believe this is going to yield the same results as removing the `scaleOutWidthHeight` call but I'm not positive. I was able to successfully launch a fullscreen window. The one warning I get is `The requested video mode is not available, switching to a valid mode`.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/compare/master...macos_fullscreen

Here's my current attempt at fixing this. It seems to work fine on my Retina Mac but am not sure how this works on non-Retina displays. If anyone else could test this branch out for me I'd really appreciate that.

**Comment by AndroFARsh** [comment]
Hello, have same issue on M1 Macbook Pro. During my research I found this [post](https://stackoverflow.com/questions/53595111/how-to-get-the-physical-display-resolution-on-macos). 

For me `CGDisplayCopyAllDisplayModes` return 60 modes when second options argument set to NULL. And there is no 1728x1117  resolution on the list, that is default on my Macbook. 
However, when I set options parameter I get 132 modes:
```
CFMutableDictionaryRef options = CFDictionaryCreateMutable(NULL, 0, &kCFCopyStringDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
CFDictionaryAddValue(options, kCGDisplayShowDuplicateLowResolutionModes, kCFBooleanTrue);
CFArrayRef cgmodes = CGDisplayCopyAllDisplayModes(CGMainDisplayID(), options);

...

CFRelease(options);
```
I don't really know how to properly fix this issue, just want to add here my investigation

**Comment by ChrisThrasher** [comment]
See https://github.com/SFML/SFML/pull/3151 for a potential fix or at least an improvement from the status quo.

**Comment by ChrisThrasher** [comment]
Fixed by #3151

**Comment by coinop-logan** [comment]
I have been watching this thread ever since I posted my summary in Jan 2023. Thank you guys so much for keeping at this!

Maybe I can build on Mac again... ❤️
